### PR TITLE
Extend ISO8601 parser to specify forbidden fields, allowing it to be used on more formats

### PR DESCRIPTION
**Pull request by thecoop**
Adds parameters to specify the maximum allowed field, decimal separator, and timezone presence, so the ISO8601 parser can be used on more built-in formats.

### CODE REVIEW

**Comment by thecoop** [comment]
@elasticmachine update branch

**Comment by thecoop** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/common/util/ArrayUtils.java [hunk 1]
```diff
@@ -68,6 +68,21 @@ public static <T> T[] concat(T[] one, T[] other) {
...
     }
 
+    /**
+     * Copy the given element and array into a new array of size {@code array.length + 1}.
+     * @param added first element in the newly created array
+     * @param array array to copy to the end of new returned array copy
+     * @return copy that contains added element and array
+     * @param <T> type of the array elements
+     */
+    public static <T> T[] append(T added, T[] array) {
```
Comment (mosche) [inline server/src/main/java/org/elasticsearch/common/util/ArrayUtils.java:78]: optional nit: maybe better `prepend`?

**Review by mosche — APPROVED** [review]
lgtm, just a minor question and a nit

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 2]
```diff
@@ -891,37 +979,75 @@ private static DateFormatter newDateFormatter(String format, DateTimeFormatter p
...
+                        false,
+                        null,
+                        DecimalSeparator.DOT,
+                        TimezonePresence.FORBIDDEN
+                    ).withLocale(Locale.ROOT),
+                    javaTimeParser }
+        );
+    }
+
+    private static final DateFormatter STRICT_DATE_HOUR_MINUTE_SECOND_MILLIS;
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java:1017]: Huh, yes they are. Hadn't spotted that. I'll look that up separately

**Pull request merged**